Purchases/Purchasing/TrialOrIntroPriceEligibilityChecker.swift

```diff
@@ -121,12 +121,59 @@ fileprivate extension TrialOrIntroPriceEligibilityChecker {
-        self.backend.getIntroEligibility(appUserID: self.appUserID,
-                                         receiptData: receiptData,
-                                         productIdentifiers: productIdentifiers) { backendResult, maybeError in
+    func getIntroEligibility(with receiptData: Data,
+                             productIdentifiers: [String],
+                             completion: @escaping ReceiveIntroEligibilityBlock) {
+        if #available(iOS 11.2, macOS 10.13.2, macCatalyst 13.0, tvOS 11.2, watchOS 6.2, *) {
+            self.productsWithIntroOffers(productIdentifiers: productIdentifiers) {
+                self.getIntroEligibility(with: receiptData,
+                                         productIdentifiers: productIdentifiers,
+                                         productIdsToIntroEligibleStatusFromApple: $0,
+                                         completion: completion)
+            }
+        } else {
+            self.getIntroEligibility(with: receiptData,
+                                     productIdentifiers: productIdentifiers,
+                                     productIdsToIntroEligibleStatusFromApple: [:],
+                                     completion: completion)
+        }
+    }
+
+    @available(iOS 11.2, macOS 10.13.2, macCatalyst 13.0, tvOS 11.2, watchOS 6.2, *)
+    func productsWithIntroOffers(productIdentifiers: [String], completion: @escaping ReceiveIntroEligibilityBlock) {
+        self.productsManager.products(withIdentifiers: Set(productIdentifiers)) { products in
+            let eligibility: [(String, IntroEligibility)] = Array(products).compactMap {
+                guard $0.introductoryPrice != nil else {
+                    return nil
+                }
+                return ($0.productIdentifier, IntroEligibility(eligibilityStatus: .eligible))
+            }
+            var productIdsToIntroEligibleStatus: [String: IntroEligibility] = [:]
+            eligibility.forEach { (productId, eligibility) in
+                productIdsToIntroEligibleStatus[productId] = eligibility
+            }
+            completion(productIdsToIntroEligibleStatus)
+        }
+    }
+
+}
+
+private extension TrialOrIntroPriceEligibilityChecker {
+
+    func getIntroEligibility(with receiptData: Data,
+                             productIdentifiers: [String],
+                             productIdsToIntroEligibleStatusFromApple: [String: IntroEligibility],
+                             completion: @escaping ReceiveIntroEligibilityBlock) {
+        // Remove any productIds we already have intro pricing for so we don't try to fetch them from the backend.
+        let idsToFetchFromBackend = productIdentifiers.filter { productIdsToIntroEligibleStatusFromApple[$0] == nil }
+        if idsToFetchFromBackend.isEmpty {
+            completion(productIdsToIntroEligibleStatusFromApple)
+            return
+        }
```

Comment:
would this be simplified if this method only received `productIdentifiers` for products that have `introductoryPrice != nil`?

Reply:
Yes, I could do the filter before this method but it made the most sense to me to do the filtering here, instead. That way any callers to this func don't have to be concerned with their own filtering.